src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -0,0 +1,1255 @@
+    // Assumption: newest unacked sent sequence number never exceeds current (and by extension, new and un-wrapped)
+    //             window boundary, so it never wraps relative to last received ack, if new window boundary would not
+    //             also wrap.
+
+    // Define new window boundary (inclusive) as uint16_t, so its value can temporarily exceed UINT8_MAX.
+    uint16_t newRemoteWindowBoundary = static_cast<uint16_t>(lastReceivedAck + maxRemoteWindowSize);
+
+    if (newRemoteWindowBoundary > UINT8_MAX && newestUnackedSentSeqNum < lastReceivedAck)
+    {
+        // New window boundary WOULD wrap, and latest unacked seq num already HAS wrapped, so add offset to difference.
+        return static_cast<uint8_t>(newRemoteWindowBoundary - (newestUnackedSentSeqNum + UINT8_MAX));
+    }
+
+    // Neither values would or have wrapped, OR new boundary WOULD wrap but latest unacked seq num does not, so no
+    // offset required.
+    return static_cast<uint8_t>(newRemoteWindowBoundary - newestUnackedSentSeqNum);
+}
+
+CHIP_ERROR WiFiPAFEndPoint::GetPktSn(Encoding::LittleEndian::Reader & reader, uint8_t * pHead, SequenceNumber_t & seqNum)
+{
+    CHIP_ERROR err;
+    BitFlags<WiFiPAFTP::HeaderFlags> rx_flags;
+    uint8_t SnOffset = 0;
+    SequenceNumber_t * pSn;
+    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    {
+        // Handkshake message => No ack/sn
+        return CHIP_ERROR_INTERNAL;
+    }
+    // Always has header flag
+    SnOffset += kTransferProtocolHeaderFlagsSize;
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
+    {
+        SnOffset += kTransferProtocolMgmtOpSize;
+    }
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    {
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn    = pHead + SnOffset;
+    seqNum = *pSn;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::DebugPktAckSn(const PktDirect_t PktDirect, Encoding::LittleEndian::Reader & reader, uint8_t * pHead)
+{
+#ifdef CHIP_WIFIPAF_END_POINT_DEBUG_LOGGING_ENABLED
+    BitFlags<WiFiPAFTP::HeaderFlags> rx_flags;
+    CHIP_ERROR err;
+    uint8_t * pAct = nullptr;
+    char AckBuff[4];
+    uint8_t * pSn;
+    uint8_t SnOffset = 0;
+
+    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    SuccessOrExit(err);
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    {
+        // Handkshake message => No ack/sn
+        return CHIP_NO_ERROR;
+    }
+    // Always has header flag
+    SnOffset += kTransferProtocolHeaderFlagsSize;
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
+    {
+        SnOffset += kTransferProtocolMgmtOpSize;
+    }
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    {
+        pAct = pHead + kTransferProtocolHeaderFlagsSize;
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn = pHead + SnOffset;
```

Comment:
How do you `pSn` is not pointing off the end of the buffer?

Reply:
The checking is required. I will add it